### PR TITLE
Checked maintainers

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
Since https://github.com/NixOS/nixpkgs/pull/66361 was merged, the maintainers list is accumulating a lot of handles without `githubId`s, or other problems such as misspelled `githubId` (partial fix in #82446) or missing name or email.

(Updated) This PR implements maintainer checks, evaluated by @GrahamcOfBorg once NixOS/ofborg#438 is merged. This guards against all of above problems. For additional convenience (and because it's possible), the error shows the correct `githubId` if it's missing:
```
lib.maintainers.infinisil: If `github` is specified, `githubId` must be too.
The GitHub ID for GitHub user infinisil is 20525370:
    githubId = 20525370;
```

All of the problems were also fixed in an initial commit, which was done by copy pasting directly from the error.

Ping @grahamc, ping @roberth for the module stuff

###### Things done

- [ ] <s>Confirmed that ofborg can catch maintainer list errors</s>

### CODE REVIEW

**Review by grahamc — CHANGES_REQUESTED** [review]
I am not yet sure we should merge this, so Requesting Changes. In short, we can't trust that the person who owned the name in the "github" is the same person who owns that name now.

When I originally filled this data in, I wrote a somewhat complicated program to look up the commit which added an entry, find the github name, see if that github name matches what is there, if so -- add the ID automatically:  https://github.com/NixOS/rfc39/blob/master/src/op_backfill.rs

I also wrote another program which verifies the author of the commit adding their maintainer entry matches the registered github ID: https://github.com/NixOS/rfc39/blob/master/src/op_blame_author.rs

On [the dashboard for the RFC-39 sync](https://status.nixos.org/grafana/d/w8HHOCBWz/rfc39-maintainer-sync?orgId=1&refresh=30s&fullscreen&panelId=30) I wrote:

```
You can't just go look up GitHub handles and assume it is right. Instead:

    Find a maintainer without a githubId
    Look in the git blame output and find that section
    Look up that commit on GitHub and see who GitHub thinks authored it.
    Find that user's ID (https://api.github.com/users/«username»)

This is because a surprising number of GitHub users change their account names!
```

Because adding their github ID to this file is the beginning of a good trust root, we should not add them lightly or casually. We will probably have a github name but not an ID for some users.

Finally, there is a question about having the maintainer list be a module itself. I am not sure that is a good idea, because some tools import that file directly and expect it to be a bit of an API. I like the validation the module gets us -- maybe that part can be in a different file, leaving this file plain data?

File: lib/options.nix
Comment (roberth) [inline lib/options.nix]: This should be renamed to `escapeNixIdentifier` and moved to `lib.strings`.

File: lib/modules.nix
Comment (roberth) [inline lib/modules.nix]: This is a huge breaking change and I don't understand why you would do this. `_module` is a perfectly valid part of `config` and being able to retrieve it is valuable, particularly when debugging.

**Review by roberth — CHANGES_REQUESTED** [review]
In addition to the comments, I'm concerned about the use of the Nix language for what appears to be data. We have perfectly usable `builtins.fromJSON` and `builtins.toJSON`, allowing anyone to write tooling that uses the data easily.

File: lib/modules.nix
Comment (edolstra) [reply lib/modules.nix]: I think the naming (starting with `_`) was intended to convey that this is not a public interface so it's fine to remove/change  it.

Comment (roberth) [reply lib/modules.nix]: It's useful for debugging with `nix repl`.

Comment (Infinisil) [reply lib/modules.nix]: Without this change:
```
$ nix-instantiate --eval -A lib.maintainers.infinisil --strict --json | jq
{
  "_module": {
    "args": {},
    "check": true
  },
  "email": "contact@infinisil.com",
  "github": "infinisil",
  "githubId": 20525370,
  "keys": [
    {
      "_module": {
        "args": {
          "name": "[definition 1-entry 1]"
        },
        "check": true
      },
      "fingerprint": "6C2B 55D4 4E04 8266 6B7D  DA1A 422E 9EDA E015 7170",
      "longkeyid": "rsa4096/0x422E9EDAE0157170"
    }
  ],
  "name": "Silvan Mosberger"
}
```

I'd have to filter out all occurrences of `_module` manually otherwise. As far as I know `_module` is only used for the module system internally.

How do you use this for debugging?

**Comment by Infinisil** [comment]
@grahamc Good point, but I think it might be okay, because from now on, as soon as people add themselves with a missing `githubId` they will get the error. This works as long as people don't change their username in the short amount of time from writing `github` to writing `githubId`.

I think the only thing to worry about is the current PRs that don't add a `githubId`. But if they get merged, master will throw the eval error, at which point we can resolve it manually until none are left anymore.

File: lib/modules.nix
Comment (roberth) [reply lib/modules.nix]: It lets me find accurately the values for the module arguments after invoking evalModules.
This is only using the module system for validation, right? Nixpkgs should use the raw data and not waste cycles on this validation. The raw data doesn't have `_module`. The validation can happen separately as part of the `lib` test suite.

**Comment by grahamc** [comment]
Note part of my concern is the IDs which are added in this PR.

File: lib/modules.nix
Comment (Infinisil) [reply lib/modules.nix]: I guess that would work in this case (but missing fields won't have a default value anymore), but in general using the raw data doesn't work, as there could be some processing that changes it, adds defaults and such. This isn't the only place where one wouldn't have to filter `_modules` anymore:
```
$ rg '"_module"'
rename.nix
17:    (mkAliasOptionModule [ "environment" "checkConfigurationOptions" ] [ "_module" "check" ])

config/krb5/default.nix
40:      (attrName: attrValue: attrName != "_module" && attrValue != null)

services/monitoring/grafana.nix
74:  _filter = x: filterAttrs (k: v: k != "_module") x;

services/monitoring/prometheus/default.nix
54:  filterValidPrometheus = filterAttrsListRecursive (n: v: !(n == "_module" || v == null));

services/networking/nghttpx/default.nix
17:      (n: v: ("_module" != n) && (null != v) && (false != v))

services/monitoring/prometheus/exporters/mail.nix
14:          (filterAttrs (n: v: !(n == "_module" || v == null)) srv)
17:    ) (filterAttrs (n: _: !(n == "_module")) cfg.configuration)
```

Note that with this change it's still possible to get `config._module` from the `config` in the argument list for modules, which can be exposed through an explicit option if wanted.

Comment (Infinisil) [reply lib/modules.nix]: E.g. we could add `_moduleArgs = mkOption { default = config._module.args; internal = true; }` in a NixOS module

Comment (roberth) [reply lib/modules.nix]: Perhaps we can return it directly from `evalModules`? So we don't have `(evalModules {}).config._module` but `(evalModules {})._module`?
It still hides the args of submodules, but that can be solved by exposing it temporarily per case.
Seems acceptable that way. 
So only these things are left iiuc.
 - Could you make a separate issue or PR for that clean-up? It's not needed for the goal of this PR.
 - This PR needs to move the validation to the test suite. (Converting to JSON is not necessary now; nice to have)
 - Graham's concern about id correctness

File: maintainers/maintainer-list.nix
Comment (FRidh) [inline maintainers/maintainer-list.nix]: This would break pure evaluation of Nixpkgs.

**Review by FRidh — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix
Comment (roberth) [reply maintainers/maintainer-list.nix]: Oh I didn't even see that code. GitHub collapses it because it's too big. Please don't mix code and data.
I recommend to check the ids without Nix. It can be made to work with fixed output derivations; it's a fun exercise, but not something to depend on.

Comment (Infinisil) [reply maintainers/maintainer-list.nix]: @FRidh nixpkgs master should never have this code be evaluating, since it will be checked before that. But yeah I guess I'll remove that, also doesn't play nicely with ofborg

File: maintainers/maintainer-list.nix
Comment (Mic92) [inline maintainers/maintainer-list.nix]: Why is it commented out?

File: maintainers/maintainer-list.nix
Comment (Mic92) [reply maintainers/maintainer-list.nix]: It needs to be checked externally anyway I suppose since github's API ratelimit kicks in after 50 requests without key.

File: maintainers/maintainer-list.nix
Comment (Infinisil) [reply maintainers/maintainer-list.nix]: This was a test to see whether ofborg would catch it (seems to have worked, but it's not very nice (also impure), so I'll change this)

File: maintainers/maintainer-list.nix
Comment (Infinisil) [reply maintainers/maintainer-list.nix]: The code here would only be executed for ones where it's missing, which after this PR should be only single digit numbers ever. But yeah that's another concern.

File: lib/modules.nix
Comment (Infinisil) [reply lib/modules.nix]: The checks are now done as part of `lib/tests/release.nix`

File: maintainers/maintainer-list.nix
Comment (Infinisil) [reply maintainers/maintainer-list.nix]: This is now being done as part of a derivation build instead.

**Comment by Infinisil** [comment]
@grahamc As discussed on IRC, I now properly verified that all GitHub IDs are correct, and yeah some did indeed change, namely:
- aminb -> @mabgnu (see https://github.com/NixOS/nixpkgs/pull/47262)
- auntie -> @auntieNeo (see https://github.com/NixOS/nixpkgs/pull/2876)
- thorstenweber83 -> @tw-360vier (see https://github.com/NixOS/nixpkgs/pull/27177, thorstenweber83 is somebody else, seems to be a mistake in https://github.com/NixOS/nixpkgs/pull/36119)

For this I used a mix of automation and manual checking. For future reference, here's a dump of my current script:

<details>

```bash
#!/usr/bin/env nix-shell
#!nix-shell -i bash -p curl jq sta

# https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line
TOKEN=
set -o pipefail

manual() {
	xdg-open "$(echo "$object" | jq -r '.url')"
	echo -n "Enter username: " >&2
	read username
	resp=$(curl -sS -H "Authorization: token $TOKEN" "https://api.github.com/users/$username")
	id=$(echo "$resp" | jq -r '.id')
	echo "For handle $handle, 'github' should be $username, 'githubId' should be $id :"
	echo "    githubId = $id;"
}

#git -C ~/src/nixpkgs log --follow --patch maintainers/maintainer-list.nix \
#  | rg -e '^index ([0-9a-f]+)\.\.([0-9a-f]+)|^commit ([0-9a-f]+)' -or '$1-$2-$3' \
#	| while IFS="-" read before after commit; do
#	  if [ -n "$commit" ]; then
#		  currentcommit="$commit"
#			echo "$currentcommit" >&2
#		else
#			echo "builtins.attrNames (let x = $(git -C ~/src/nixpkgs show "$before"); in if builtins.isFunction x then x {} else x)" > beforefile
#			echo "builtins.attrNames (let x = $(git -C ~/src/nixpkgs show "$after"); in if builtins.isFunction x then x {} else x)" > afterfile
#		  if ! keysbefore=$(nix-instantiate beforefile --eval --json --strict | jq '.[]' -r); then
#			  continue
#			fi
#		  if ! keysafter=$(nix-instantiate afterfile --eval --json --strict | jq '.[]' -r); then
#			  continue
#			fi
#			diff \
#				--new-line-format="%L" \
#				--old-line-format="" \
#				--unchanged-line-format="" \
#				<(echo "$keysbefore") <(echo "$keysafter") | while read user; do
#				echo "Handle $user was added in $currentcommit" >&2
#				echo "$user $currentcommit"
#			done
#
#		fi
#	done
#
#exit 0

#response=$(curl -sS -H "Authorization: token $TOKEN" https://api.github.com/users/$1)
#
#echo "    githubId = $(echo "$response" | jq -r '.id');" | tee /dev/stderr |  /nix/store/5ghdbn3lmbaqdj7dba0a5m1hb2949zwy-xclip-0.13/bin/xclip -selection clipboard
#exit 0
forHandle() {
  local handle=$1
  local github=$2
	echo -n "For handle $handle the previous github login was $github. " >&2
	commit=$(rg "^$handle (.*)$" -or '$1' maintainer-map | tail -1)
	#commit=$(git -C ~/src/nixpkgs log --follow -G "$\s*$handle[^a-zA-Z0-9_'-]^" --pretty=format:%H maintainers/maintainer-list.nix | tail -1)
	echo -n "Found commit $commit . " >&2

	query=$(cat <<-EOF
{
  repository(name: "nixpkgs", owner: "NixOS") {
    object(expression: "$commit") {
      ... on Commit {
        associatedPullRequests(first: 10) {
          nodes {
            author {
              ... on User {
                databaseId
                login
              }
            }
          }
        }
        author {
          user {
            databaseId
            login
          }
        }
				url
      }
    }
  }
}
	EOF
	)
	DATA=$(jq -n --arg query "$query" '{ query : $query }')
	response=$(curl -sS -H "Authorization: token $TOKEN" -d "$DATA" https://api.github.com/graphql)
	object=$(echo "$response" | jq '.data.repository.object')
  result=$(echo "$object" | jq '[ ( .author.user , .associatedPullRequests.nodes[].author) ] | map(select(. != null)) | unique[] | "\(.login) \(.databaseId)"' -r)
	if [ -z "$result" ]; then
	  count=0
	else
		count=$(echo "$result" | wc -l)
	fi
	if [ "$count" -eq 0 ]; then
		echo -e "\e[31mCouldn't figure out who the author is!\e[0m" >&2
		manual
	elif [ "$count" -eq 1 ]; then
		read login id <<< "$result"
		echo -n "Found author with login $login and id $id. " >&2
		if [ "$github" = "$login" ]; then
		  echo -e "\e[32mThis is the same as the one in the maintainer file\e[0m" >&2
			echo "For handle $handle, 'github' should be $login, 'githubId' should be $id :"
			echo "    githubId = $id;"
		else
		  echo -e "\e[31mThis is NOT the same as the one in the maintainer file!\e[0m" >&2
			manual
		fi
	else
		echo -e "\e[31mDifferent commit author than PR author!\e[0m" >&2
		manual
	fi
}

#forHandle DerGuteMoritz DerGuteMoritz
#forHandle Fresheyeball fresheyeball
#forHandle MP2E MP2E
#forHandle Mogria mogria
#forHandle MtP MtP76
#forHandle abbradar abbradar
#forHandle acairncross acairncross
#forHandle adamt adamtulinius
#forHandle aforemny aforemny
#forHandle ak alexanderkjeldaas
#forHandle alunduil alunduil
#forHandle ambrop72 ambrop72
#forHandle aminb aminb
#forHandle andreabedini andreabedini
#forHandle andres kosmikus
#forHandle andsild andsild
#forHandle antono antono
#forHandle auntie auntie
#forHandle averelld averelld
#forHandle backuitist backuitist
#forHandle badi badi
#forHandle berdario berdario
#forHandle bluescreen303 bluescreen303
#forHandle boothead boothead
#forHandle coconnor coreyoconnor
#forHandle cransom cransom
#forHandle davidrusu davidrusu
#forHandle desiderius desiderius
#forHandle dgonyeo dgonyeo
#forHandle doublec doublec
#forHandle dxf dingxiangfei2009
#forHandle edanaher edanaher
#forHandle emmanuelrosa emmanuelrosa
#forHandle ericsagnes ericsagnes
#forHandle ertes ertes
#forHandle fare fare
#forHandle fdns fdns
#forHandle fragamus fragamus
#forHandle freezeboy freezeboy
#forHandle garbas garbas
#forHandle gavin gavinrogers
#forHandle gridaphobe gridaphobe
#forHandle hkjn hkjn
#forHandle infinisil infinisil
#forHandle j03 johannesloetzsch
#forHandle jasoncarr jasoncarr0
#forHandle jeschli jeschli
#forHandle jfb tftio
#forHandle jitwit jitwit
#forHandle joamaki joamaki
#forHandle joepie91 joepie91
#forHandle jonathanmarler marler8997
#forHandle juliendehos juliendehos
#forHandle jyp jyp
#forHandle kampfschlaefer kampfschlaefer
#forHandle knairda KnairdA
#forHandle koral k0ral
#forHandle laikq laikq
#forHandle lassulus Lassulus
#forHandle lebastr lebastr
#forHandle leonardoce leonardoce
#forHandle lovek323 lovek323
#forHandle ltavard ltavard
#forHandle lumi lumi-me-not
#forHandle matthewbauer matthewbauer
#forHandle matti-kariluoma matti-kariluoma
#forHandle melsigl melsigl
#forHandle michaelpj michaelpj
#forHandle michelk michelk
#forHandle minijackson minijackson
#forHandle mirdhyn mirdhyn
#forHandle mkf mkf
#forHandle mmlb mmlb
#forHandle mpscholten mpscholten
#forHandle mredaelli mredaelli
#forHandle nand0p nand0p
#forHandle ngerstle ngerstle
#forHandle olynch olynch
#forHandle orbitz orbitz
#forHandle pcarrier pcarrier
#forHandle plchldr plchldr
#forHandle pmeunier P-E-Meunier
#forHandle polyrod polyrod
forHandle rafaelgg rafaelgg
#forHandle rickynils rickynils
#forHandle rob rbvermaa
#forHandle robberer robberer
#forHandle rvolosatovs rvolosatovs
#forHandle ryansydnor ryansydnor
#forHandle sander svanderburg
#forHandle scalavision scalavision
#forHandle schmittlauch schmittlauch
#forHandle scubed2 scubed2
#forHandle shazow shazow
#forHandle shlevy shlevy
#forHandle shmish111 shmish111
#forHandle sjmackenzie sjmackenzie
#forHandle sprock sprock
#forHandle srghma srghma
#forHandle taha tgharib
#forHandle tckmn tckmn
#forHandle tesq0 tesq0
#forHandle teto teto
#forHandle the-kenny the-kenny
#forHandle timbertson timbertson
#forHandle timma ktrsoft
#forHandle tnias tnias
#forHandle tscholak tscholak
#forHandle tvestelind tvestelind
#forHandle tweber thorstenweber83
#forHandle twey twey
#forHandle uwap uwap
#forHandle valeriangalliat valeriangalliat
#forHandle vcanadi vcanadi
#forHandle viric viric
#forHandle vizanto vizanto
#forHandle vmchale vmchale
#forHandle wscott wscott
#forHandle xnaveira xnaveira
#forHandle y0no y0no
#forHandle zalakain umazalakain
```
</details>

The tests are now also run as part of `lib/tests/release.nix` and are separate of the maintainer file. They are now implemented as a fixed-output derivation.

**Comment by Infinisil** [comment]
@grahamc Looks like ofborg doesn't run/build lib tests (only instantiate them via nixpkgs-tarball), which is needed for this to work now. See https://github.com/NixOS/ofborg/pull/438 which should add this.

**Comment by thorstenweber83** [comment]
thorstenweber83 and tw-360vier are indeed the same person. I have a separate work account :/
Please use thorstenweber83 for the maintainers file.
I'll add a comment with tw-360vier in a moment.

**Comment by tw-360vier** [comment]
Please use thorstenweber83 for the maintainers file :)

**Review by Ma27 — APPROVED** [review]

**Comment by Infinisil** [comment]
Note that there's also some maintainer checks in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/misc/meta.nix . Ideally that should be combined with this

**Comment by Infinisil** [comment]
GitHub just had a bad outage (see https://www.githubstatus.com/incidents/80d0cs6kpsps) right when I pushed an update to this PR, which apparently somehow caused a whole lot of commits to be included (which were already in master), causing a lot of reviews to be requested.

File: lib/strings.nix
Comment (rycee) [inline lib/strings.nix:326]: Minor:
```suggestion
       => "\"0abc\""
```

**Comment by Infinisil** [comment]
Checks are working nicely now that https://github.com/NixOS/ofborg/pull/438 is merged :D. See https://gist.github.com/GrahamcOfBorg/c999ed1b03e7061cdd4f213cac6e394d for how such failures look. I'll now update the PR to fix those two instances of people not setting `githubId`s. Once merged, we'll probably have to fix those a bunch more times in the near future as old PRs get merged.

File: lib/tests/maintainers.nix
Comment (rycee) [inline lib/tests/maintainers.nix:77]: Since the `info` variable is unused I guess one could assign directly to `id` using some piping:
```suggestion
        id=$(curl -sS https://api.github.com/users/${checkedAttrs.github} | jq -r '.id')
```

Comment (Infinisil) [reply lib/tests/maintainers.nix:77]: I've had to debug this a bit, which I did with `echo "$info"` inbetween those lines. I think this might come in handy in the future, and it keeps the line width down too.

Comment (rycee) [reply lib/tests/maintainers.nix:77]: Fair enough. I think it overall looks very nice. Good job!

**Review by rycee — APPROVED** [review]

**Review by roberth — APPROVED** [review]

**Review by Ma27 — APPROVED** [review]
Great work!

**Comment by Infinisil** [comment]
Rebased on top of master to fix another case of a missing `githubId`

**Comment by Infinisil** [comment]
This evening I'll fix up any new potential missing `githubId`s and then merge it myself if there aren't any complaints.

**Comment by Infinisil** [comment]
Turns out I forgot about doing it that evening. Done it now though, only one `githubId` had to be fixed since then. One other change I made is that I removed the `description`s from the options, because they're useless and would be duplicated (once in `lib/tests/maintainers.nix` and once in `maintainers/maintainer-list.nix`).

I'll merge once @GrahamcOfBorg is happy

**Comment by Infinisil** [comment]
PR that reverts that commit is now merged: https://github.com/NixOS/nixpkgs/pull/85241

**Pull request merged**
